cmd/lotus-shed/splitstore.go

```diff
@@ -0,0 +1,238 @@
+		}
+		defer lr.Close() //nolint:errcheck
+
+		cfg, err := lr.Config()
+		if err != nil {
+			return xerrors.Errorf("error getting config: %w", err)
+		}
+
+		fncfg, ok := cfg.(*config.FullNode)
+		if !ok {
+			return xerrors.Errorf("wrong config type: %T", cfg)
+		}
+
+		if !fncfg.Chainstore.EnableSplitstore {
+			return xerrors.Errorf("splitstore is not enabled")
+		}
+
+		fmt.Println("copying hotstore to coldstore...")
+		err = copyHotstoreToColdstore(lr)
+		if err != nil {
+			return xerrors.Errorf("error copying hotstore to coldstore: %w", err)
+		}
+
+		fmt.Println("deleting splistore directory...")
+		err = deleteSplitstoreDir(lr)
+		if err != nil {
+			return xerrors.Errorf("error deleting splitstore directory: %w", err)
+		}
+
+		fmt.Println("deleting splitstore keys from metadata datastore...")
+		err = deleteSplitstoreKeys(lr)
+		if err != nil {
+			return xerrors.Errorf("error deleting splitstore keys: %w", err)
+		}
+
+		fmt.Println("disabling splitstore in config...")
+		err = lr.SetConfig(func(cfg interface{}) {
+			cfg.(*config.FullNode).Chainstore.EnableSplitstore = false
+		})
+		if err != nil {
+			return xerrors.Errorf("error disabling splitstore in config: %w", err)
+		}
+
+		fmt.Println("splitstore has been rolled back.")
+		return nil
+	},
+}
+
+func copyHotstoreToColdstore(lr repo.LockedRepo) error {
+	repoPath := lr.Path()
+	dataPath := filepath.Join(repoPath, "datastore")
+	coldPath := filepath.Join(dataPath, "chain")
+	hotPath := filepath.Join(dataPath, "splitstore", "hot.badger")
+
+	coldOpts, err := repo.BadgerBlockstoreOptions(repo.UniversalBlockstore, coldPath, false)
+	if err != nil {
+		return xerrors.Errorf("error getting coldstore badger options: %w", err)
+	}
+	coldOpts.SyncWrites = false
+
+	hotOpts, err := repo.BadgerBlockstoreOptions(repo.HotBlockstore, hotPath, true)
+	if err != nil {
+		return xerrors.Errorf("error getting hotstore badger options: %w", err)
+	}
+
+	cold, err := badger.Open(coldOpts.Options)
+	if err != nil {
+		return xerrors.Errorf("error opening coldstore: %w", err)
+	}
+	defer cold.Close() //nolint
+
+	hot, err := badger.Open(hotOpts.Options)
+	if err != nil {
+		return xerrors.Errorf("error opening hotstore: %w", err)
+	}
+	defer hot.Close() //nolint
+
+	rd, wr := io.Pipe()
+	g := new(errgroup.Group)
+
+	g.Go(func() error {
+		bwr := bufio.NewWriterSize(wr, 64<<20)
+
+		_, err := hot.Backup(bwr, 0)
+		if err != nil {
+			_ = wr.CloseWithError(err)
+			return err
+		}
+
+		err = bwr.Flush()
+		if err != nil {
+			_ = wr.CloseWithError(err)
+			return err
+		}
+
+		return wr.Close()
+	})
+
+	g.Go(func() error {
+		err := cold.Load(rd, 1024)
+		if err != nil {
+			return err
+		}
+
+		return cold.Sync()
+	})
+
+	err = g.Wait()
+	if err != nil {
+		return err
+	}
+
+	// compact + gc the coldstore
+	fmt.Println("compacting coldstore...")
+	nworkers := runtime.NumCPU()
+	if nworkers < 2 {
+		nworkers = 2
+	}
+
+	err = cold.Flatten(nworkers)
```

Comment:
How long can this take? Because if it takes more than a couple minutes, it's probably best to have it as an option as it's not functionally necessary.

Reply:
no, it shouldn't take more than a couple of minutes normally but it might for big coldstores.
So yeah, I'll put it behind an option.

Reply:
done.